### PR TITLE
[Block Editor]: Conditionally show `Styles` if block is in Placeholder state

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Partially addresses: https://github.com/WordPress/gutenberg/issues/23122 
Related: https://github.com/WordPress/gutenberg/pull/33466

There is the need to show/hide some controls or other things like `Styles` based on whether the block is in `Placeholder` state or not. For example when we have an `Image` block with no image set, there is no need to show the `Styles` tab or the `Styles` in `BlockSwitcher`.

This PR adds to the `block-editor` store an `inPlaceholderState` which tracks which blocks are in such state. This value can be consumed to conditionally show/hide things. In order to be a complete solution for `block.supports` it will need one more property to the state per `support`.

For start I've implemented this for `Image` block and will wait for feedback for the approach. If there is a consensus, I'll follow up on this by writing tests and add it to more blocks that can utilize this like Cover block(issue: https://github.com/WordPress/gutenberg/issues/23198)

## Testing instructions
1. Insert an `Image` block and observe that if you have not set an image, the `Styles` are not shown.
2. Set an image and observe that `Styles` are shown.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by youknowriad** [comment]
Thanks for the PR Nik. It sounds like a valid approach but I'd like to take some time to reflect on what it means more holistically.

Right now, block objects can have a `isValid` boolean. Which basically means a block can be in two states: "valid" or "invalid". This PR adds two new states for the block "placeholder" and "ready/customizable". We could see these as two completely different flags or states like this PR does. Alternatively we could see as "states" of a block. The question that I have:

 - Is the isPlaceholder state a formal state we want to introduce for all blocks?
 - Is there any other block state we think we'd need in the future? What other states the block can be in?
 - Should these states be part of the block object (like isValid) or not like (just in the block-editor store) the isPlaceholder of this PR? (I think I prefer the latter here)
 - Should these states and any potential new states have a generic implementation (block states) and a state machine.

Some questions raised by this PR, they don't need to be answered or solved all in this PR, just food for thoughts.

cc @mcsf @mtias

**Comment by Mamaduka** [comment]
Thanks for the PR, @ntsekouras.

It never occurred to me that we can store placeholder state this way. 🙇 

My ideas always leaned towards `block.supports`.

* Blocks can define support for placeholder state and provide an array of attributes for this property.
* The `isInPlaceholderState` value is derived based on these attributes.

```json
{
	"apiVersion": 2,
	"name": "core/image",
	"attributes": {},
	"supports": {
		"anchor": true,
		"placeholder": [ "url" ]
	}
}
```

**Comment by oandregal** [comment]
I've left a comment about this at https://github.com/WordPress/gutenberg/pull/33466#issuecomment-891784831 I think something like this is our best shot, although it still has drawbacks for third-party tools.

**Comment by ajlende** [comment]
@youknowriad I know your questions were meant to be rhetorical, but I've been thinking about this for a bit now, so I have some answers for some of them anyway.

> Is the isPlaceholder state a formal state we want to introduce for all blocks?

I wasn't sure about this either, but I asked during the [#core-editor chat](https://wordpress.slack.com/archives/C02QB2JS7/p1626879601454100) and others seemed to think that it is. Looking around, more than a few blocks already have some from of a setup state where nothing will be displayed unless some attributes are set. In these cases, the block usually doesn't want to give a full set of controls until those required attributes are set. 

> Should these states and any potential new states have a generic implementation (block states) and a state machine?

One of my ideas would fit well with a state machine type implementation.

Given that setup state is a thing that a block developer wants to use for their block (probably not too uncommon based on the above), from a block developer perspective, what would the easiest way to handle a setup state be?

One of my ideas is adding a `setup` or `placeholder` function alongside `edit` and `save`. Block supports would be able to decide if they should be visible in the placeholder state or not. I haven't dug too deeply into block styles yet, but they could be marked either when registering the style or disabled altogether when in the setup state. The challenge is deciding how to transition from the setup to the edit state. Would we need to supply another function, call it `isSetupComplete` or something, for that transition, or would a simple check of `prop !== undefined` be enough?

**Comment by ntsekouras** [comment]
Thanks for the feedback all!

> Is the isPlaceholder state a formal state we want to introduce for all blocks?

I believe we could - there are many blocks that use Placeholders. Also, with the state approach we leave blocks (objects) intact. 

>Is there any other block state we think we'd need in the future? What other states the block can be in?

I'm not sure, but probably an `editable/read only - locked` state in the future, as is being discussed in this issue: https://github.com/WordPress/gutenberg/issues/29864. Having this info in the state could be consumed by components to show/hide things in toolbar/list view etc.. 

>Should these states and any potential new states have a generic implementation (block states) and a state machine.

If we don't find/agree on another possible state for now, I'd be reluctant to implement a generic solution. I think it would be better to test it out as `experimental`. Although probably the `locked` state is something we will need 😄 

>Blocks can define support for placeholder state and provide an array of attributes for this property.

@Mamaduka this can easily fall short for many blocks that need more complex calculation for this state. For example, lots of blocks are in Placeholder state when there are no `innerBlocks`. 

>One of my ideas is adding a setup or placeholder function alongside edit and save.

@ajlende since a `Placeholder` can be different in blocks and is tied to the `edit` function, IMO is okay to set this state in the same function.

**Comment by ajlende** [comment]
> since a Placeholder can be different in blocks and is tied to the edit function

I'm questioning if placehoders between blocks are different _enough_? And if it _has_ to be tied to the edit function?

What would be easier for a block developer to understand? Having a separate function for when you have required props that need to be set or adding additional wiring and branching within the edit function and having two different returns in one function? (I don't really know for sure)

Having two different returns is what we already have—it's the more flexible solution, and it's able to do the things that we want to do. However, the additional wiring (the `useEffect` calls) can be hidden by having a separate block state for the common case of having some required setup. Maybe that would be easier for block developers?

And I just thought of this, but we want the edit function to be visually as close to what is saved, right? In the case of things like images where we've completely replaced the UI in this setup case, having a separate setup function could help support that goal too.

**Comment by gziolo** [comment]
There are many core blocks that use the placeholder state. Let me share some examples so we have a better understanding of their complexity.

## Columns block

https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/columns/edit.js#L271-L280

## Query Loop block

https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/query/edit/index.js#L157-L163

## Gallery block

https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/gallery/edit.js#L383
https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/gallery/edit.js#L412-L414

## RSS block

https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/rss/edit.js#L28
https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/rss/edit.js#L59-L82

## Table block

https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/table/edit.js#L343-L345
https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/table/edit.js#L425
https://github.com/WordPress/gutenberg/blob/62b80f9794f6e6334303253d3aaa3dc53fb40aa4/packages/block-library/src/table/edit.js#L503-L538

**Comment by jasmussen** [comment]
> I'm questioning if placehoders between blocks are different enough? And if it has to be tied to the edit function?

I'm speaking here from a pure design and user experience perspective, and I might be missing some nuance of the development side of things, so what I'm about to suggest might or might not be completely relevant.

Nevertheless the `Placeholder`s as they exist today (white bordered boxes) have a number of shortcomings that we need to address in the future:

- They look nothing like the end result. They don't pass the "squint" test.
- Because of that, they often cause big layout shifts when going from the placeholder state to the final state. 
- For example, the Placeholder comes with a baked-in 200px min-height, which is not at all appropriate for a horizontal Navigation menu, for example, as those are usually 36px tall. [Here's an older example of why that approach isn't ideal](https://user-images.githubusercontent.com/1204802/80569553-7326b500-89f9-11ea-8dff-4c196b252df6.jpg).
- It works best when inserted in a big column; if you insert it in a very small context, although there's some limited responsive behavior built in, it's not very good in the default state. Each placeholder requires careful designs and configurations in order to behave well to the context it's inserted in.

In many blocks, the `Placeholder` is critical for the block to function. An Image block needs an image chosen, a particular Maps block might need an API key, an embed block needs a URL. In those cases, the component and "Windows 95 Wizard-like" behavior can be a good way to configure a complex block. But even here, the shortcomings noted above are still present. 

In other cases, we use the Placeholder component even though we arguably shouldn't, for all the reasons listed above. For example the Table block allows you to configure the number of columns and rows in the Placeholder itself, but arguably we could just default to 2x2 and then let you change those from the block toolbar directly. This isn't meant as a critique of the choice, often the Placeholder was chosen because the block toolbar wasn't yet a good enough user experience. But it needs to be, and we need fewer of those Placeholders.

In short, what we've learned from the user experience side is that designing a good block setup state requires careful consideration of its responsive behavior, consideration of whether it's even necessary in the first place, and thoughts about how to best pass the "squint" test.

We _might_ still be able to create some smart behaviors that handle these things in a mostly generic way, or at least provide props that help attune the component to its block type. I've revisited a few recently, though important to note I'm still not entirely happy with these:

Setup states   appenders

Setup states   appenders 2

Generic placeholder

But all that is to say: the Placeholder is not as simple as it might seem, and is likely going to vary a great deal from block to block.

**Comment by ajlende** [comment]
Thanks @gziolo and @jasmussen for the examples!

> There are many core blocks that use the placeholder state. Let me share some examples so we have a better understanding of their complexity.

It seems that the placeholders are simpler than I expected—driven almost entirely by attributes and things tied to the block clientId. Only RSS requires internal block state, and that could easily be avoided by entering the url in the toolbar like the image url in the replacement flow currently is done.

image replacement flow

> But all that is to say: the Placeholder is not as simple as it might seem, and is likely going to vary a great deal from block to block.

Yeah, the design of the Placeholder is not simple—I don't think we could narrow down to even a few categories of placeholder to render. However, it seems that choosing _if_ a block should be in a placeholder state is pretty simple—it's not so simple that we could encode it in JSON, but a single function seems like it could work.

**Comment by ntsekouras** [comment]
I'm closing this for now, as this seems to be subjective about user's workflow. Might be better to address the shortcomings and usage of Placeholders in the first place, as explained very well by @jasmussen in the above comment:https://github.com/WordPress/gutenberg/pull/33823#issuecomment-892443035
